rust_src/src/multibyte.rs

```diff
@@ -42,32 +43,285 @@ use crate::{
+        }
+    }
+
+    // Equivalent to ASCII_CHAR_P
+    /// Whether the character is an ASCII value
+    pub const fn is_ascii(self) -> bool {
+        self.0 < MIN_2_BYTE_CHAR
+    }
+
+    // Equivalent to SINGLE_BYTE_CHAR_P
+    /// Whether the character is a single byte value (i.e. less than 256/0x100)
+    pub const fn is_single_byte(self) -> bool {
+        self.0 < 0x100
+    }
+
+    // Equivalent to CHAR_BYTE8_P
+    /// Whether the character is a byte8 value, i.e. an encoded raw 8bit byte.
+    pub const fn is_byte8(self) -> bool {
+        self.0 >= MIN_BYTE8_CHAR
+    }
+
+    /// The amount of bytes needed to represent the codepoint's multibyte form.
+    pub fn len_bytes(self) -> usize {
+        match self.0 {
+            0..=MAX_1_BYTE_CHAR => 1,
+            MIN_2_BYTE_CHAR..=MAX_2_BYTE_CHAR => 2,
+            MIN_3_BYTE_CHAR..=MAX_3_BYTE_CHAR => 3,
+            MIN_4_BYTE_CHAR..=MAX_4_BYTE_CHAR => 4,
+            MIN_5_BYTE_CHAR..=MAX_5_BYTE_CHAR => 5,
+            MIN_BYTE8_CHAR..=MAX_BYTE8_CHAR => 2,
+            _ => invalid_character(self),
+        }
+    }
+
+    /// Get the integer value of the codepoint
+    pub const fn val(self) -> u32 {
+        self.0
+    }
+
+    // Equivalent to CHAR_TO_BYTE8
+    /// Extract the encoded byte8 or raw byte value.
+    ///
+    /// Note that this does not check if the codepoint is within the
+    /// appropriate range.
+    pub fn to_byte8_unchecked(self) -> u8 {
+        match self.is_byte8() {
+            true => (self.0 - BYTE8_OFFSET) as u8,
+            false => (self.0 & 0xFF) as u8,
+        }
+    }
+
+    // Equivalent to CHAR_TO_BYTE_SAFE
+    /// Return the raw 8-bit byte for the character, or None if it doesn't
+    /// correnspond to a byte.
+    pub fn to_byte8(self) -> Option<u8> {
+        if self.is_ascii() {
+            Some(self.0 as u8)
+        } else if self.is_byte8() {
+            Some((self.0 - BYTE8_OFFSET) as u8)
+        } else {
+            None
+        }
+    }
+
+    // Equivalent to UNIBYTE_TO_CHAR
+    pub fn unibyte_to_char(self) -> Codepoint {
+        match self.is_ascii() {
+            true => self,
+            false => Codepoint::from_raw(self.0 as u8),
+        }
+    }
+
+    // Equivalent to MAKE_CHAR_MULTIBYTE
+    /// Transform an 8-bit codepoint to its byte8 encoded form.
+    pub fn to_multibyte(self) -> Codepoint {
+        debug_assert!(self.is_single_byte());
+        self.unibyte_to_char()
+    }
+
+    // Equivalent to CHAR_STRING
+    /// Write the codepoint to the given slice. The slice needs to be big
+    /// enough to hold the resulting bytes. Returns the amount of bytes written
+    pub fn write_to(self, to: &mut [u8]) -> usize {
+        let cp: u32 = self.into();
+        if cp <= MAX_1_BYTE_CHAR.into() {
+            to[0] = cp as u8;
+            1
+        } else if cp <= MAX_2_BYTE_CHAR {
+            // note: setting later bytes first to avoid multiple bound checks
```

Comment:
+1

Reply:
although you might move the comment up a level, since it applies to most of the branches.

Reply:
+1 as in "this is good" or "change the 2 to a 3"?